nexus/src/app/rack.rs

```diff
@@ -933,33 +935,76 @@ impl super::Nexus {
     }
 
     /// Add a sled to an intialized rack
-    pub(crate) async fn add_sled_to_initialized_rack(
+    pub(crate) async fn sled_add(
         &self,
         opctx: &OpContext,
-        sled: UninitializedSled,
+        sled: UninitializedSledId,
     ) -> Result<(), Error> {
-        let baseboard_id = sled.baseboard.clone().into();
-        let hw_baseboard_id =
-            self.db_datastore.find_hw_baseboard_id(opctx, baseboard_id).await?;
+        let baseboard_id = sled.clone().into();
+        let hw_baseboard_id = self
+            .db_datastore
+            .find_hw_baseboard_id(opctx, &baseboard_id)
+            .await?;
 
-        let subnet = self.db_datastore.rack_subnet(opctx, sled.rack_id).await?;
+        let subnet = self.db_datastore.rack_subnet(opctx, self.rack_id).await?;
         let rack_subnet =
             Ipv6Subnet::<RACK_PREFIX>::from(rack_subnet(Some(subnet))?);
 
         let allocation = self
             .db_datastore
             .allocate_sled_underlay_subnet_octets(
                 opctx,
-                sled.rack_id,
+                self.rack_id,
                 hw_baseboard_id,
             )
             .await?;
 
+        // Grab the SPs from the last collection
+        //
+        // This is necessary to get the revision_id for the given sled
+        //
+        // We set a limit of 200 here to give us some breathing room when
+        // querying for cabooses and RoT pages, each of which is "4 per SP/RoT",
+        // which in a single fully populated rack works out to (32 sleds + 2
+        // switches + 1 psc) * 4 = 140.
+        //
+        // This feels bad and probably needs more thought; see
+        // https://github.com/oxidecomputer/omicron/issues/4621 where this limit
+        // being too low bit us, and it will link to a more general followup
+        // issue.
+        let limit = NonZeroU32::new(200).unwrap();
+        let collection = self
+            .db_datastore
+            .inventory_get_latest_collection(opctx, limit)
+            .await?;
+
+        // If there isn't a collection, we don't know about the sled
+        let Some(collection) = collection else {
+            return Err(Error::ObjectNotFound {
+                type_name: ResourceType::Sled,
+                lookup_type:
+                    omicron_common::api::external::LookupType::ByCompositeId(
+                        format!("{sled:?}"),
+                    ),
+            });
```

Comment:
I think I might make this a 500 or 503 with internal message "no inventory data available".  The problem here doesn't seem like it's that they gave us a sled that doesn't exist -- it's that something is seriously broken because we have no inventory data at all.

Reply:
Fixed in [d14ca51](https://github.com/oxidecomputer/omicron/pull/4704/commits/d14ca5157c6a7eb130b613a057042f011a1a1fb8)